### PR TITLE
Add `observation.metadata.choice_nodes`

### PR DESCRIPTION
**Pull request by tybug**
Closes (most of) https://github.com/HypothesisWorks/hypothesis/issues/4387

- [x] depends on https://github.com/HypothesisWorks/hypothesis/pull/4416

To do:
- [x] document the `choice_nodes` format (nonstandard nan patterns, large ints, etc)
- [x] document `observability.OBSERVABILITY_CHOICE_NODES`
- [ ] work out a proper `@settings` for this? (as a follow-up?)

~~I'm also combining this with changes to `BuildContext` (for hypofuzz), so that this PR can be "the entirety of the changes necessary to make HypofuzzProvider work".~~ I've split this out and have dropped it from this PR

### CODE REVIEW

File: hypothesis-python/RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+RELEASE_TYPE: minor
+
+This release adds |OBSERVABILITY_CHOICE_NODES|, a new option for :ref:`observability <observability>`, which includes the choice sequence in ``metadata.choice_nodes`` for test case observations if set.
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst]: I think the changelog should explicitly note that this is an experimental/unstable feature, and thus also be for a change release (since there's no change to our _stable_ API)

File: hypothesis-python/tests/cover/test_observability.py [hunk 2]
```diff
@@ -332,3 +340,152 @@ def test_fails(should_fail, should_fail_assume):
     assert len(ls) == 1
     assert ls[0].status == expected_status
     assert ls[0].how_generated == "fuzz_one_input"
+
+
+def _decode_choice(value):
+    if isinstance(value, list):
+        if value[0] == "integer":
+            # large integers get cast to string, stored as ["integer", str(value)]
+            assert isinstance(value[1], float)
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_observability.py]:
```suggestion
            # large integers get cast to string, stored as ["integer", str(value)]
            assert isinstance(value[1], str)
```
right?  (if so, needs more test)

File: hypothesis-python/tests/cover/test_observability.py [hunk 3]
```diff
@@ -332,3 +340,152 @@ def test_fails(should_fail, should_fail_assume):
...
+@example([math.inf])
+@example([-math.inf])
+@given(st.lists(choices()))
+def test_choices_json_roundtrips(choices):
+    # choices_to_json and nodes_to_json roundtrip, *except for large integers*,
+    # which get cast to the nearest integer-valued float on roundtrip. This is
+    # an intentional design decision of the format; see related comment in
+    # to_jsonable.
+    if any(_will_be_cast_to_float(choice) for choice in choices):
+        assume(False)
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_observability.py]: `st.lists(choices().filter(lambda x: not _will_be_cast_to_float(x))` would be much more efficient; and the same below.  I'd also reverse the sense of this, and define `_expect_to_roundtrip()` instead.

File: hypothesis-python/docs/reference/schema_metadata.json [hunk 4]
```diff
@@ -0,0 +1,91 @@
+{
+    "title": "Hypothesis Metadata",
```
Comment (Zac-HD) [inline hypothesis-python/docs/reference/schema_metadata.json]: This renders as a second title in the docs; consider removing or excluding?

File: hypothesis-python/docs/reference/schema_metadata.json [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            "enum": [0, 1, 2, 3],
+            "description": "The internal status of the ConjectureData for this test case. The values are as follows: ``Status.OVERRUN = 0``, ``Status.INVALID = 1``, ``Status.VALID = 2``, and ``Status.INTERESTING = 3``."
+        },
+        "interesting_origin": {
+            "type": ["string", "null"],
+            "description": "The internal InterestingOrigin object for failing tests, if and only if ``status == \"failed\"``. The ``traceback`` string value is derived from this object."
+        },
+        "choice_nodes": {
+            "type": ["array", "null"],
+            "description": "The sequence of choices made during this test case. This includes the choice value, as well as its constraints and whether it was forced or not. The choice sequence is a relatively low-level implementation detail of Hypothesis, and is exposed here for users building tools or research on top of Hypothesis. See |PrimitiveProvider| for more details about the choice sequence.\n\nOnly present if |OBSERVABILITY_CHOICE_NODES| is set.",
```
Comment (Zac-HD) [inline hypothesis-python/docs/reference/schema_metadata.json]: again, I think we need to specifically note that this is unstable / experimental API.

File: hypothesis-python/docs/reference/schema_metadata.json [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        "data_status": {
+            "type": "number",
+            "enum": [0, 1, 2, 3],
+            "description": "The internal status of the ConjectureData for this test case. The values are as follows: ``Status.OVERRUN = 0``, ``Status.INVALID = 1``, ``Status.VALID = 2``, and ``Status.INTERESTING = 3``."
+        },
+        "interesting_origin": {
+            "type": ["string", "null"],
+            "description": "The internal InterestingOrigin object for failing tests, if and only if ``status == \"failed\"``. The ``traceback`` string value is derived from this object."
+        },
```
Comment (Zac-HD) [inline hypothesis-python/docs/reference/schema_metadata.json]: IIRC we already store this in the top-level `status_reason` key.

[goes and checks] aaaaaannnd, we mishandle non-failing cases, oops.

File: hypothesis-python/src/hypothesis/internal/observability.py [hunk 7]
```diff
@@ -231,6 +356,15 @@ def _system_metadata() -> dict[str, Any]:
...
     "HYPOTHESIS_EXPERIMENTAL_OBSERVABILITY_NOCOVER" not in os.environ
 )
+#: If ``True``, include the ``metadata.choice_nodes`` key in test case
+#: observations.
+#:
+#: ``False`` by default. ``metadata.choice_nodes`` can be substantial amount of
+#: data, and so must be opted-in to, even when observability is enabled.
+OBSERVABILITY_CHOICE_NODES = (
+    "HYPOTHESIS_EXPERIMENTAL_OBSERVABILITY_CHOICE_NODES" in os.environ
+)
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/observability.py]: also needs the experimental / unstable note, precisely because I'm keen to find a better interface for this.

**Review by Zac-HD — COMMENTED** [review]
Do we have a concrete use-case in mind for the node constraints?  If not, how about b64-encoded bytes format for the choice sequence?  It'd be a fair bit smaller, and it looks like the human-readable version is already complicated enough that outside parties will want to use our code to deal with _that_.

Either way, I was thinking that we could have a flat list of `[start, end, was_discarded, label]` values, derived from the `Spans` class to track the tree structure of the input.

File: hypothesis-python/docs/reference/schema_metadata.json (hunk 6 above)
Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/schema_metadata.json]: yeah, INVALID and OVERRUN get collapsed together (which is probably-good as a library agnostic format)

**Comment by Liam-DeVoe** [comment]
> Do we have a concrete use-case in mind for the node constraints? If not, how about b64-encoded bytes format for the choice sequence? It'd be a fair bit smaller, and it looks like the human-readable version is already complicated enough that outside parties will want to use our code to deal with that.

Concrete use case is passing in-memory to hypofuzz, mostly for ordering corpus elements, cmd+f "choice_nodes" in https://github.com/Zac-HD/hypofuzz/pull/113/files. 

I don't have a consumer of the serialized file format in mind, but my thinking is "if you're in a position to be using hypothesis utilities to transform the observations, you should just hook `OBSERVABILITY_CALLBACKS` and work with it in-memory (or serialize it yourself in your own format to a file). And if you're not in a position to use hypothesis utilities (working in another language etc), you'll greatly benefit from a human-readable format than having to replicate our `choices_from_bytes` function"

File: hypothesis-python/docs/reference/schema_metadata.json (hunk 6 above)
Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/schema_metadata.json]: sorry, I thought this was `status`, not `status_reason`. The actual reason I split this out is because hypofuzz wants the `InterestingOrigin` object itself, *not* the str(origin) version in `status_reason`.

[e: and this means the type here is wrong, fixing..]

[e2: nope, it is a string because we're documenting the file format here, not the in-memory format. Hmm, unfortunate that the two are similar but different...]

**Comment by Zac-HD** [comment]
Yeah, OK, fair enough.  Feels inelegant but I think moving fast at the cost of inelegant formats is the right call for now.

File: hypothesis-python/tests/cover/test_observability.py (hunk 2 above)
Comment (Liam-DeVoe) [reply hypothesis-python/tests/cover/test_observability.py]: right - this was from when I was using `float(n)` instead of the current `str(n`); the latter roundtrips and the former does not. Filtering out non-roundtripping-values was missing this case. I've just dropped that filtering since we expect everything to roundtrip now.

**Comment by Liam-DeVoe** [comment]
> [start, end, was_discarded, label

I went with `[str(span.label), span.start, span.end, span.discarded]`

File: hypothesis-python/src/hypothesis/internal/observability.py [hunk 8]
```diff
@@ -139,6 +284,10 @@ def make_testcase(
     from hypothesis.core import reproduction_decorator
     from hypothesis.internal.conjecture.data import Status
 
+    # We should only be sending observability reports for datas that have finished
+    # being modified.
+    assert data.frozen
+
     if status_reason is not None:
```
Comment (Liam-DeVoe) [inline hypothesis-python/src/hypothesis/internal/observability.py:291]: `data` needs to be frozen to access` data.spans`, and at that point it seems reasonable to require that all datas passed here are frozen. I don't think freezing more than we used to at any of the callers is harmful, and anyway the whole point of `.freeze` is a phase-transition at which point it cannot be modified, which we definitely expect if we're reporting an observation for a data.

File: hypothesis-python/RELEASE.rst [hunk 9]
```diff
@@ -0,0 +1,5 @@
+RELEASE_TYPE: patch
+
+This release adds the experimental and unstable |OBSERVABILITY_CHOICES| option for :ref:`observability <observability>`, which includes the choice sequence in ``metadata.choice_nodes`` for test case observations if set.
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst]: also `choice_spans`!

File: hypothesis-python/docs/reference/integrations.rst [hunk 10]
```diff
@@ -162,6 +162,15 @@ including Gson in Java, ``JSON.parse()`` in Ruby, and of course in Python.
    :hide_key: /additionalProperties, /type
 
 
+Hypothesis Metadata
+^^^^^^^^^^^^^^^^^^^
```
Comment (Zac-HD) [inline hypothesis-python/docs/reference/integrations.rst]: Let's

- move the information-message schema first
- make this a subheading of the test-case schema
- consider pulling out the choice_nodes and choice_spans to a second subheading and collapsing `<details>` by default (it's a _lot_ of very-niche-interest text)

File: hypothesis-python/docs/reference/schema_metadata.json [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                "additionalProperties": false
+            }
+        },
+        "choice_spans": {
+            "type": "array",
+            "items": {"type": "array"},
+            "description": ".. warning::\n\n  EXPERIMENTAL AND UNSTABLE. This attribute may change format or disappear without warning.\n\nThe semantically-meaningful spans of the choice sequence of this test case.\n\nEach span has the format ``[label, start, end, discarded]``, where:\n\n* ``label`` is an opaque integer-value string shared by all spans drawn from a particular strategy.\n* ``start`` and ``end`` are indices into the choice sequence for this span, such that ``choices[start:end]`` are the corresponding choices.\n* ``discarded`` is a boolean indicating whether this span was discarded (see |PrimitiveProvider.span_end|).\n\nOnly present if |OBSERVABILITY_CHOICES| is ``True``.\n\n.. note::\n\n  Spans are a relatively low-level implementation detail of Hypothesis, and are exposed in observability for users building tools or research on top of Hypothesis. See |PrimitiveProvider| (and particularly |PrimitiveProvider.span_start| and |PrimitiveProvider.span_end|) for more details about spans."
```
Comment (Zac-HD) [inline hypothesis-python/docs/reference/schema_metadata.json]: ~~We should also say explicitly that the label -> strategy mapping may be different in each test case; it's just "the `n`th strategy".~~

Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/schema_metadata.json]: really, it can be different? is that only in edge cases or dynamic construction? it's pretty hard to do any kind of research on top of spans if they aren't stable across test cases

Comment (Zac-HD) [reply hypothesis-python/docs/reference/schema_metadata.json]: Oops, no, when I dig through the code I was thinking of something else.  In general they _can_ vary between runs ([e.g. `calc_label_from_hash`](https://github.com/HypothesisWorks/hypothesis/blob/80669745b130e2fd8dd1ce09d0761d40fa499432/hypothesis-python/src/hypothesis/internal/conjecture/utils.py#L32-L42)), [see also `calc_label`](https://github.com/HypothesisWorks/hypothesis/blob/80669745b130e2fd8dd1ce09d0761d40fa499432/hypothesis-python/src/hypothesis/strategies/_internal/strategies.py#L579-L626), but should be stable between test cases for a single process.

Which probably still needs an explanatory note!

edit: I'd also be tempted to represent it in observability as a b64 string, since they're morally speaking hashes.

File: hypothesis-python/docs/reference/integrations.rst (hunk 10 above)
Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/integrations.rst]: I'm not sure sphinx has collapsible elements unless we install some third party extension 🤔

Comment (Zac-HD) [reply hypothesis-python/docs/reference/integrations.rst]: It's a cute trick, the built-in `only` directive allows for arbitrary html:

```rst
.. only:: html

   .. raw:: html

      <details>
      <summary>Click to expand</summary>

Put whatever you want here, and in the HTML build only it'll be inside a details tag!

.. only:: html

   .. raw:: html

      </details>
```

Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/integrations.rst]: happy to collapse for now, but TBH I think observability might deserve its own page at this point. (I think ghostwriter does as well). IMO a selling point of big api reference pages is that you don't have to worry about things like niche content or collapsing text 

~~(which also makes cmd+f harder)~~ TIL cmd+f autoexpands `<details>`

but still, collapsing makes it easy to miss, and this is a pretty important section *for the right subset of people*

Comment (Zac-HD) [reply hypothesis-python/docs/reference/integrations.rst]: no, good argument, leave it uncollapsed and we'll move it later.

File: hypothesis-python/docs/reference/schema_metadata.json (hunk 11 above)
Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/schema_metadata.json]: re edit: tempting, but makes it harder/more confusing for any researcher combining the in-memory integer label and the b64 label from serialized observations

Comment (Liam-DeVoe) [reply hypothesis-python/docs/reference/schema_metadata.json]: let me know if you still want the new text to be changed; I don't really want to say "labels may change between test cases" because that's not really true, it's only for dynamically-constructed strategies like changing what you pass to sampled_from, at which point I would call that a distinct strategy

**Review by Zac-HD — APPROVED** [review]
